File: week9_inception/solutions/inception/python/nq.native.py
import sys

def print_solution(board):
    """Helper function to print the board configuration."""
    n = len(board)
    for row in board:
        print(" ".join('Q' if i == row else '.' for i in range(n)))
    print("-----")

def solve_nqueens(n):
    """Function to solve the N-Queens problem."""
    board = [-1] * n  # board[i] = the column of the queen in the ith row
    solutions = []

    def is_safe(row, col):
        """Check if it's safe to place a queen at (row, col)."""
        for i in range(row):
            if board[i] == col or abs(board[i] - col) == row - i:
                return False
        return True

    def solve(row):
        """Recursive function to solve the N-Queens problem."""
        if row == n:
            solutions.append(board[:])  # Found a solution
            return
        for col in range(n):
            if is_safe(row, col):
                board[row] = col
                solve(row + 1)
                board[row] = -1  # Backtrack

    solve(0)
    return solutions

def main():
    """Main function to handle input and display solutions."""
    if len(sys.argv) != 2:
        print("Usage: python3 nqueens.py <n>")
        sys.exit(1)

    try:
        n = int(sys.argv[1])
    except ValueError:
        print("Please provide a valid integer for n.")
        sys.exit(1)

    if n <= 0:
        print("Please provide a valid positive integer for n.")
        sys.exit(1)

    print(f"Solving N-Queens for n = {n}...")
    solutions = solve_nqueens(n)

    if solutions:
        print(f"Found {len(solutions)} solution(s):")
        for solution in solutions:
            print_solution(solution)
    else:
        print("No solution exists.")

if __name__ == "__main__":
    main()

File: challenges/week9_inception/solutions/inception/go/nq.native.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"math"
)

var count int

// solve is the backtracking function for solving N-Queens problem
func solve(n, col int, hist []int) {
	if col == n {
		// Print the solution
		count++
		fmt.Printf("\nNo. %d\n-----\n", count)
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				if j == hist[i] {
					fmt.Print("Q ")
				} else {
					// Print alternating spaces and dots
					if (i+j)%2 == 1 {
						fmt.Print("  ")
					} else {
						fmt.Print(". ")
					}
				}
			}
			fmt.Println()
		}
		return
	}

	// attack checks if the queen in row i and column col is attacked
	for i := 0; i < n; i++ {
		// Check if placing the queen in row i, column col is safe
		attack := false
		for j := 0; j < col; j++ {
			if hist[j] == i || int(math.Abs(float64(hist[j]-i))) == col-j {
				attack = true
				break
			}
		}
		if attack {
			continue
		}

		// Place queen in hist[col] = i (row i, column col)
		hist[col] = i
		// Recurse to solve for next column
		solve(n, col+1, hist)
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Please provide the number of queens as a command-line argument.")
		return
	}

	// Parse the number of queens from command-line argument
	n, err := strconv.Atoi(os.Args[1])
	if err != nil || n <= 0 {
		fmt.Println("Invalid input. Please provide a positive integer for the number of queens.")
		return
	}

	// Initialize the histogram to hold the positions of queens in each column
	hist := make([]int, n)

	// Start solving the N-Queens problem
	fmt.Printf("Solving %d-Queens Problem:\n", n)
	solve(n, 0, hist)
}
